cpu/cortexm_common/thread_arch.c

```diff
@@ -296,84 +308,110 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
 
     /* skip context saving if sched_active_thread == NULL */
     "ldr    r1, =sched_active_thread  \n" /* r1 = &sched_active_thread  */
-    "ldr    r1, [r1]                  \n" /* r1 = sched_active_thread   */
-    "cmp    r1, #0                    \n" /* if r1 == NULL:             */
-    "beq select_and_restore_context   \n" /*   goto select_and_restore_context */
+#if CPU_CORE_CORTEX_FULL_THUMB
+    "ldr    r12, [r1]                 \n" /* r12 = sched_active_thread   */
+#else
+    "ldr    r1, [r1]                  \n"
+    "mov    r12, r1                   \n" /* r12 = sched_active_thread   */
+#endif
+    "push   {lr}                      \n" /* push exception return code */
+
+    "bl     sched_run                 \n" /* perform scheduling */
+
+#if CPU_CORE_CORTEX_FULL_THUMB
+    "cmp    r0, r12                   \n" /* if r0 == 0: (no switch required) */
+    "it     eq                        \n"
+    "popeq  {pc}                      \n" /* Pop exception to pc to return */
+
+    "pop    {lr}                      \n" /* Pop exception from the exception stack */
+
+    "mov    r1,r12                    \n" /* r1 = sched_active_thread */
+    "cbz    r1, restore_context       \n" /* goto restore_context if r1 == 0 */
+
+    "mrs    r2, psp                   \n" /* get stack pointer from user mode */
+
+#if (defined(CPU_CORE_CORTEX_M4F) || defined(CPU_CORE_CORTEX_M7)) && defined(MODULE_CORTEXM_FPU)
```

Comment:
Why is this only needed on Cortex-M4F and Cortex-M7?
That is, why not just check for `MODULE_CORTEXM_FPU`?
Is Cortex-M33 saving the registers automatically?

Reply:
I've simplied copied it over from [here](https://github.com/RIOT-OS/RIOT/pull/14710/files#diff-ceb0a6b8e226f3f54bf29e05b6d8a7d1L322). I know that this is not written with cortex-m33 in mind and honestly I consider that out of scope here. I can fix the check if you have a proposal, but that's going to be a separate PR :)

Reply:
Why not simply
```C
#if defined(MODULE_CORTEXM_FPU)
```

`MODULE_CORTEXM_FPU`  *can't* be defined without either of the two other symbols also being defined, given that this was written with only `CPU_CORE_CORTEX_M4F` and `CPU_CORE_CORTEX_M7` in mind.